src/core/friendly_errors/param_validator.js

```diff
@@ -8,351 +8,366 @@ import { z } from 'zod';
+   *
+   * Example parameter data for function `background`:
+   * "background": {
+        "overloads": [
+          ["p5.Color"],
+          ["String", "Number?"],
+          ["Number", "Number?"],
+          ["Number", "Number", "Number", "Number?"],
+          ["Number[]"],
+          ["p5.Image", "Number?"]
+        ]
+      }
+   * Where each array in `overloads` represents a set of valid overloaded
+   * parameters, and `?` is a shorthand for `Optional`.
+   *
+   * TODO:
+   * - [ ] Support for p5 constructors
+   * - [ ] Support for more obscure types, such as `lerpPalette` and optional
+   * objects in `p5.Geometry.computeNormals()`
+   * (see https://github.com/processing/p5.js/pull/7186#discussion_r1724983249)
+   *
+   * @param {String} func - Name of the function.
+   * @returns {z.ZodSchema} Zod schema
+   */
+  function generateZodSchemasForFunc(func) {
+    // Expect global functions like `sin` and class methods like `p5.Vector.add`
+    const ichDot = func.lastIndexOf('.');
+    const funcName = func.slice(ichDot + 1);
+    const funcClass = func.slice(0, ichDot !== -1 ? ichDot : 0) || 'p5';
+
+    let funcInfo = arrDoc[funcClass][funcName];
+
+    let overloads = [];
+    if (funcInfo.hasOwnProperty('overloads')) {
+      overloads = funcInfo.overloads;
     }
 
-    const requiredParamsCount = params.filter(p => !p.endsWith('?')).length;
-    const result = [];
-
-    for (let i = requiredParamsCount; i <= params.length; i++) {
-      result.push(params.slice(0, i));
-    }
+    // Returns a schema for a single type, i.e. z.boolean() for `boolean`.
+    const generateTypeSchema = type => {
+      // Type only contains uppercase letters and underscores -> type is a
+      // constant. Note that because we're ultimately interested in the value of
+      // the constant, mapping constants to their values via `constantsMap` is
+      // necessary.
+      if (/^[A-Z_]+$/.test(type)) {
+        return z.literal(constantsMap[type]);
+      }
+      // All p5 objects start with `p5` in the documentation, i.e. `p5.Camera`.
+      else if (type.startsWith('p5')) {
+        console.log('type', type);
+        const className = type.substring(type.indexOf('.') + 1);
+        console.log('className', p5Constructors[className]);
+        return z.instanceof(p5Constructors[className]);
+      }
+      // For primitive types and web API objects.
+      else if (schemaMap[type]) {
+        return schemaMap[type];
+      } else {
+        // TODO: Make this throw an error once more types are supported.
+        console.log(`Warning: Zod schema not found for type '${type}'. Skip mapping`);
+        return undefined;
+      }
+    };
 
-    return result;
-  };
+    // Generate a schema for a single parameter. In the case where a parameter can
+    // be of multiple types, `generateTypeSchema` is called for each type.
+    const generateParamSchema = param => {
+      const optional = param.endsWith('?');
```

Comment:
It might be a good idea to handle arrays like this too when generating a param schema by checking for something that ends with `[]` to save having to manually handle array versions of each type

Reply:
Sounds good!

Reply:
Just pushed another commit that handles `[]` logic dynamically!